### PR TITLE
created early CLI prototype

### LINKED ISSUE #275: Setup an initial database for local development
**Is your feature request related to a problem?**

This feature is to ease a developer's life when running the app for simple testing, or to test a specific feature.

I want to have a way to setup an initial database, when I run the project. If I run the project I don't want to have always to setup the entire dataset, I want the option to use an existing database.

Example: I want to clone run the project with the option to use dataset 1 with 3 users, 2 of them in a mentorship relation with a couple of tasks. Once this is loaded, when I run the project, I can test the local app knowing I have these already created.

**Describe the solution you'd like**

Not sure about the solution to be honest. Open for suggestions!

**Describe alternatives you've considered**

Broadly speaking... have a local environment variable to say if a data is to be loaded, and have a part of the code, such as feature toggle, that if that is set a specific dataset is loaded.

Be aware that, for user emails, this can be configured at anytime. Imagining i'm using temporary emails, not everyday I will have the same emails to work with, so emails should not be "set in stone".

**Additional context**

Maybe this can be solved in a way with #78 ... I'm not sure, but its worth mentioning.

_If anyone needs more clarification I'd be more than happy to help with this_

### PR DESCRIPTION
**Pull request by bartekpacia**
### Description
In my opinion, working with the backend using Postman isn't really a great experience. `POST` Register user with a real email, open mailbox and confirm, then `POST` login. And if you want to use many other endpoints, you also need JWT for a particular user. And that user sometimes needs to be an admin (when u perform admin-like things). Huh, a lot of boring and tedious work.

I've seen issue #275, then #78, and thought...let's do it!
I decided to go with [Click](https://click.palletsprojects.com/en/7.x/), as it's provided by Flask out-of-the-box (Flask uses it internally for its own stuff, too). It's also very cool.
From an "architectural side" I use [Blueprints](https://flask.palletsprojects.com/en/1.0.x/blueprints/). They fit this use case quite well and are fairly simple.

And here it is, an early CLI prototype. I really hope it makes its way to the merge with main branch. It will save a lot of time:)

Features:
- dropping and creating database
- populating database with initial data
- creating, updating and removing users
- easy listing of all users and specific user (by id, username and email)
- listing, assigning and revoking admins
- managing mentorship relations
- managing tasks

To be added:
- unit tests
- some decent documentation

fixes #275, #78 (well, it *will fix* when it gets stable)

### Type of Change:
- Code
- Quality Assurance
- User Interface [I'm not sure here...CLI is kinda like UI for a developer)
- Documentation

I think it's worth mentioning that I had to upgrade `Flask` dependency to `1.1.1`. That's beacuse [this very useful functionality](https://github.com/pallets/flask/issues/1357) is available since `1.1.0`. And `1.1.0` messed some things that `1.1.1` fortunately fixed.

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)

### How Has This Been Tested?
Tested manually, but as said above, this is an early prototype. Consider this whole things as unstable - no unit tests, functionality isn't guarranteed to work. Use at you own risk (in the worst case you'll nuke your local database).

### How to use it? (Example)
1. Make sure that virtualenv is activated.
2. `flask` (yea, just type `flask` in the terminal)
3. You'll see commands at the bottom
4. Let's say you want to create a new user.
5. `flask user` will show you methods available for managing users.
6. You see that `flask user create` is what you need. Type it in the terminal.
7. Follow the instructions on screen.
8. User is added.
9. Now let's display all users using `flask user show -s`. -s flag shows only essential data.

I encourage you to explore and try (and find inevitable bugs). I'll try to write some more examples or maybe even ~~record a short video tutorial on how to use it~~  [demo/tutorial here](https://www.youtube.com/watch?v=W5ZRI88wg5U)

### Checklist:
- [ ] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas (I will)
- [ ] I have made corresponding changes to the documentation (I will)
- [ ] Any dependent changes have been merged
- [x] Update requirements.txt

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

PS Any feedback would be greatly appreciated. I'm coming from NodeJS-land and still feel quite new and uncertain in this astonishing Python-Flasky world.

PPS I'll squash the commits when we it gets approved by mentors.

### CODE REVIEW

**Comment by bartekpacia** [comment]
@isabelcosta I think it may interest you

**Review by isabelcosta — COMMENTED** [review]
Thank you for tackling this problem @bartekpacia !This is not  code review, per say, but a general comment for now. As the issue here is not part of GCI task, then this PR should be targeting `develop` branch.

**Comment by bartekpacia** [comment]
Sure, changed :)

**Comment by bartekpacia** [comment]
I've recorded [video demonstration](https://www.youtube.com/watch?v=W5ZRI88wg5U).

**Comment by bartekpacia** [comment]
@isabelcosta @BadduCoder I'd love to hear your opinion and suggestions 😄 (if you have time of course)

**Comment by isabelcosta** [comment]
@bartekpacia This is my first time looking and actually thinking about this, so I just watched your video and this is pretty amazing :) And I just love that this helps to create a set of users just from a JSON file.
I feel that this will require some changes, I will review soon. Do you wish to continue with this issue, if I end up proposing changes?

I already learned new things from this PR.

**Comment by bartekpacia** [comment]
@isabelcosta Oh, this PR. I remember how much work I put into this:) happy to hear you like it. I'd love this to be developed further because it has the potential of saving many hours of rather tedious work (I remember I was using this very often during GCI for my private testing purposes).

I'll wait for your opinions and will try to find some time to polish this. Hopefully it'll make its way into master.

**Comment by bartekpacia** [comment]
i accidentaly closed👀

*help*

**Comment by isabelcosta** [comment]
I accidentally opened 👀

**Review by Hyper-Matrix — DISMISSED** [review]

**Comment by robotjellyzone** [comment]
hi @bartekpacia could you mind to resolve the conflicts in this PR so, that it can be ready to merge :)

**Comment by bartekpacia** [comment]
@robotjellyzone Sure, resolving now.

**Comment by bartekpacia** [comment]
I've spent 1 hour trying to setup dev env for this project, but some extremely strange bugs appear.
Finally, I resolved it using GitHub's website editor, so I couldn't test it :/

**Comment by bartekpacia** [comment]
at least checks are passing🙏

**Comment by bartekpacia** [comment]
@isabelcosta any response on this? I'd really love to see this get merged.
(It might need some testing & updates, but I'm willing to provide necessary help)

**Comment by bartekpacia** [comment]
just a ping @isabelcosta :)

File: run.py [hunk 1]
```diff
@@ -41,9 +52,14 @@ def create_app(config_filename: str) -> Flask:
 @application.before_first_request
 def create_tables():
     from app.database.sqlalchemy_extension import db
-
     db.create_all()
 
 
 if __name__ == "__main__":
     application.run(port=5000)
+
```
Comment (isabelcosta) [inline run.py:60]: can you add a comment here, to explain you are adding cli options or something, also should this be inside or outside the create_app()?

File: run.py [hunk 2]
```diff
@@ -1,7 +1,17 @@
...
+import cli.admin
+import cli.database
+import cli.mentorship_relation
+import cli.task
+import cli.user
+from app.database.sqlalchemy_extension import db
 from config import get_env_config
 from flask_migrate import Migrate
 
+load_dotenv(verbose=True)
```
Comment (isabelcosta) [inline run.py:13]: is this really needed?

**Review by isabelcosta — CHANGES_REQUESTED** [review]
Ok after a quick look at this, here's what I would suggest for this PR to be better:
- Add instructions on how to work with this tool (commands, and one example per command)on a markdown file under `docs` folder. In the end link it in the README somehow.
- solve merge conflicts (e.g.: we are now using flask-restx instead of flask-restplus)
- if possible, add unit tests for the cli options you created
- (nice to have) add typehints :)

@bartekpacia thank you for chasing me to review this 🙈 , this seems great work and to ease developer's work, without affecting functionalities of the project.

**Comment by vj-codes** [comment]
Closing this due to inacativity also author is not assigned to the issue .

**Comment by bartekpacia** [comment]
@vj-codes , I'd really like to finish this feature:) Please consider reopening and assigning me – I will try to work on it when I have time. 
Are there any hard limits on the duration of how long can a PR be worked on by an author?

**Comment by vj-codes** [comment]
@bartekpacia suree I'll assign it to you!
No limits on duration but you should update the progress or approx time required so the PR is not inactive otherwise it could be closed by someone again 👀

**Comment by vj-codes** [comment]
@bartekpacia also there are some merge conflicts do solve them when you have time, thanks!

**Comment by bartekpacia** [comment]
That's great, thank you:)

**Comment by bartekpacia** [comment]
I'm looking at the list of all my PRs and this is one of the oldest ones. Unfortunately I don't see a chance on getting this merged, last change (d878a013060785a0a077ea16342707b501880827) was made more than a year ago, so I'm going to reluctantly close this PR. Maybe, hopefully, in the future someone will be willing to pick this up.

Good old Google Code-in times. Thanks to everyone involved!

**Pull request closed without merging**